File: src/main/java/com/EmosewaPixel/pixellib/materialSystem/types/BlockType.java
package com.EmosewaPixel.pixellib.materialsystem.types;

import com.EmosewaPixel.pixellib.materialsystem.MaterialRegistry;
import com.EmosewaPixel.pixellib.materialsystem.materials.IMaterialItem;
import com.EmosewaPixel.pixellib.materialsystem.materials.Material;
import com.google.gson.JsonObject;
import net.minecraft.block.Block;

import java.util.function.Function;
import java.util.function.Predicate;

//Block Types are Object Types used for generating Blocks
public class BlockType extends ObjectType {
    private Block.Properties properties;
    private Function<IMaterialItem, JsonObject> blockstateJson;

    public BlockType(String name, Predicate<Material> requirement, Block.Properties properties) {
        super(name, requirement);
        this.properties = properties;
        blockstateJson = i -> {
            JsonObject states = new JsonObject();
            JsonObject variants = new JsonObject();
            JsonObject variant = new JsonObject();
            variant.addProperty("model", "pixellib:block/materialblocks/" + (hasTag(MaterialRegistry.SINGLE_TEXTURE_TYPE) ? getName() : i.getMat().getTextureType().toString() + "/" + getName()));
            variants.add("", variant);
            states.add("variants", variants);
            return states;
        };
    }

    public BlockType setBlockstateJson(Function<IMaterialItem, JsonObject> blockstateJson) {
        this.blockstateJson = blockstateJson;
        return this;
    }

    public Block.Properties getProperties() {
        return properties;
    }

    public JsonObject getBlockstateJson(IMaterialItem item) {
        return blockstateJson.apply(item);
    }
}

File: src/main/java/com/emosewapixel/pixellib/materialsystem/types/BlockType.kt
package com.emosewapixel.pixellib.materialsystem.types

import com.blamejared.crafttweaker.api.annotations.ZenRegister
import com.emosewapixel.pixellib.blocks.MaterialBlock
import com.emosewapixel.pixellib.blocks.MaterialBlockItem
import com.emosewapixel.pixellib.extensions.json
import com.emosewapixel.pixellib.materialsystem.addition.MaterialRegistry
import com.emosewapixel.pixellib.materialsystem.materials.IMaterialObject
import com.emosewapixel.pixellib.materialsystem.materials.Material
import com.emosewapixel.pixellib.materialsystem.materials.utility.ct.MaterialRequirement
import com.google.gson.JsonObject
import net.minecraft.block.Block
import net.minecraft.item.Item
import net.minecraft.util.text.TranslationTextComponent
import org.openzen.zencode.java.ZenCodeType

//Block Types are Object Types used for generating Blocks
@ZenRegister
@ZenCodeType.Name("pixellib.materialsystem.types.BlockType")
class BlockType @JvmOverloads @ZenCodeType.Constructor constructor(name: String, requirement: MaterialRequirement, val properties: Block.Properties, constructor: (Material, BlockType) -> Block = ::MaterialBlock, val itemConstructor: (Material, BlockType) -> Item = ::MaterialBlockItem) : ObjectType<Block, BlockType>(name, requirement, constructor) {
    var buildBlockStateJson: (IMaterialObject) -> JsonObject = {
        json {
            "variants" {
                "" {
                    "model" to "pixellib:block/materialblocks/" + if (MaterialRegistry.SINGLE_TEXTURE_TYPE in typeTags) name else "${it.mat.textureType}/$name"
                }
            }
        }
    }

    override fun localize(mat: Material) = TranslationTextComponent("blocktype.$name", mat.localizedName)
}